From 50a11303a622479d6164d3dd2ea8ec6a2c6f2658 Mon Sep 17 00:00:00 2001
From: ywan171 <yi.a.wang@intel.com>
Date: Mon, 6 May 2013 22:08:22 +0800
Subject: libstagefright: add protection when connection and disconnect http in different thread

BZ: 75893

orig-patch-id: 84819,94891

Category: aosp improvement
Domain: Video.Media-streaming
Origin: internal
Upstream-Candidate: yes

Change-Id: I27e42459828467f0682bb546cf95e8c14228206e
Orig-Change-Id: I95d376f86a8decbebb88dd7dc563c17639e0790d
Signed-off-by: ywan171 <yi.a.wang@intel.com>
Signed-off-by: chuzhoux <chux.zhou@intel.com>
---
 .../chromium_http/ChromiumHTTPDataSource.cpp       |    9 +++++++++
 .../include/ChromiumHTTPDataSource.h               |    2 ++
 2 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/chromium_http/ChromiumHTTPDataSource.cpp b/media/libstagefright/chromium_http/ChromiumHTTPDataSource.cpp
index a862d8b..ab4571e 100644
--- a/media/libstagefright/chromium_http/ChromiumHTTPDataSource.cpp
+++ b/media/libstagefright/chromium_http/ChromiumHTTPDataSource.cpp
@@ -37,6 +37,7 @@ ChromiumHTTPDataSource::ChromiumHTTPDataSource(uint32_t flags)
       mIOResult(OK),
       mContentSize(-1),
       mDecryptHandle(NULL),
+      mDisconnectExternal(false),
       mDrmManagerClient(NULL) {
     mDelegate->setOwner(this);
 }
@@ -86,6 +87,12 @@ status_t ChromiumHTTPDataSource::connect_l(
                 "connect to <URL suppressed> @%lld", offset);
 #endif
 
+    if (mDisconnectExternal) {
+        LOG_PRI(ANDROID_LOG_INFO, LOG_TAG,
+               "should not connect when disconnect is already called in another thread");
+	return mIOResult;
+    }
+
     mURI = uri;
     mContentType = String8("application/octet-stream");
 
@@ -136,7 +143,9 @@ void ChromiumHTTPDataSource::onConnectionFailed(status_t err) {
 
 void ChromiumHTTPDataSource::disconnect() {
     Mutex::Autolock autoLock(mLock);
+    mDisconnectExternal = true;
     disconnect_l();
+    mDisconnectExternal = false;
 }
 
 void ChromiumHTTPDataSource::disconnect_l() {
diff --git a/media/libstagefright/include/ChromiumHTTPDataSource.h b/media/libstagefright/include/ChromiumHTTPDataSource.h
index 785f939..4ccd72c 100644
--- a/media/libstagefright/include/ChromiumHTTPDataSource.h
+++ b/media/libstagefright/include/ChromiumHTTPDataSource.h
@@ -90,6 +90,8 @@ private:
 
     int64_t mContentSize;
 
+    bool mDisconnectExternal;
+
     String8 mContentType;
 
     sp<DecryptHandle> mDecryptHandle;
-- 
1.7.4.1

